ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -252,42 +268,68 @@ initRangeReadLMDBTable count (LMDBMK _        db)                      = ApplyVa
 rangeReadLMDBTable     count (LMDBMK _        db) (ApplyKeysMK   keys) = ApplyValuesMK    <$> lmdbRangeReadTable count db keys
 
 {-------------------------------------------------------------------------------
- Db settings
+ Db state
 -------------------------------------------------------------------------------}
 
-readDbSettingsMaybeNull ::
+readDbStateMaybeNull ::
      LMDB.Database () DbState
   -> LMDB.Transaction mode (Maybe DbState)
-readDbSettingsMaybeNull db = LMDB.get db ()
+readDbStateMaybeNull db = LMDB.get db ()
 
-withDbSettingsRW ::
+withDbStateRW ::
      LMDB.Database () DbState
   -> (DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState))
   -> LMDB.Transaction LMDB.ReadWrite a
-withDbSettingsRW db f = withDbSettingsRWMaybeNull db $ maybe (Exn.throw DbErrNoSettings) f
+withDbStateRW db f = withDbStateRWMaybeNull db $ maybe (Exn.throw DbErrNoDbState) f
 
-withDbSettingsRWMaybeNull ::
+withDbStateRWMaybeNull ::
       LMDB.Database () DbState
    -> (Maybe DbState -> LMDB.Transaction LMDB.ReadWrite (a, DbState))
    -> LMDB.Transaction LMDB.ReadWrite a
-withDbSettingsRWMaybeNull db f  =
-  readDbSettingsMaybeNull db >>= f >>= \(r, new_s) -> LMDB.put db () (Just new_s) $> r
+withDbStateRWMaybeNull db f  =
+  readDbStateMaybeNull db >>= f >>= \(r, new_s) -> LMDB.put db () (Just new_s) $> r
 
 data GuardDbDir  = GDDMustExist | GDDMustNotExist
 
-guardDbDir :: MonadIO m => GuardDbDir -> FS.SomeHasFS m -> FS.FsPath -> m FilePath
+-- | Guard for the existence/non-existence of a database directory,
+-- and create it if missing.
+guardDbDir ::
+     (MonadIO m, IOLike m)
+  => GuardDbDir
+  -> FS.SomeHasFS m
+  -> FS.FsPath
+  -> m FilePath
 guardDbDir gdd (FS.SomeHasFS fs) path = do
-  FS.doesFileExist fs path >>= \b -> when b $ Exn.throw $ DbErrStr $ "guardDbDir:must be a directory:" <> show path
-  FS.doesDirectoryExist fs path >>= \case
-    True | GDDMustNotExist <- gdd ->
-           -- TODO Should throw a DbErrDbExists exception
-           -- Callers should delete the directory if they want to restore a snapshot
-           FS.removeDirectoryRecursive fs path
-    False | GDDMustExist <- gdd -> Exn.throw $ DbErrDbDoesntExist path
-    _ -> pure ()
+  fileEx <- FS.doesFileExist fs path
+  when fileEx $
+    Exn.throw $ DbErrStr $ "guardDbDir:must be a directory: " <> show path
+  dirEx <- FS.doesDirectoryExist fs path
+  case dirEx of
+    True  | GDDMustNotExist <- gdd -> throwIO $ DbErrDbExists path
+    False | GDDMustExist    <- gdd -> throwIO $ DbErrDbDoesntExist path
+    _                              -> pure ()
   FS.createDirectoryIfMissing fs True path
   FS.unsafeToFilePath fs path
 
+-- | Same as @`guardDbDir`@, but retries the guard if we can make meaningful
+-- changes to the filesystem before we perform the retry.
+--
+-- Note: We only retry if a database directory exists while it shoudn't: In
```

Comment:
I wonder if this is what we want. Don't we risk deleting a database just because, for instance, we passed the wrong flags?